src/dinitctl.cc

```diff
@@ -1822,12 +1822,44 @@ static int enable_disable_service(int socknum, cpbuffer_t &rbuffer, service_dir_
         }
     }
 
+    bool wait_for_event = true;
     if (socknum >= 0) {
         // warn if 'from' service is not started
-        if (enable && from_state != service_state_t::STARTED) {
-            cerr << "dinitctl: warning: enabling dependency for non-started service" << endl;
-        }
+        if (enable) {
+            if (from_state != service_state_t::STARTED) {
+                cerr << "dinitctl: warning: enabling dependency for non-started service" << endl;
+            }
+
+            // Check the service status before sending ENABLESERVICE to avoid a
+            // "SERVICESTATUS vs SERVICEEVENT" race condition.
+            auto m = membuf()
+                    .append((char)cp_cmd::SERVICESTATUS)
+                    .append(to_handle);
+            write_all_x(socknum, m);
```

Comment:
This introduces another race. If you check the service state before sending ENABLESERVICE, the state can change in between the check and sending ENABLESERVICE.

The best fix is probably to not use SERVICESTATUS at all, at least when enabling a service, and rely on the status information (event) packets. For disabling a service SERVICESTATUS is necessary as there is no guarantee a service will stop when it is disabled and we want to warn in case there is a separate dependency that holds it started.

Reply:
> To avoid the race you would need to send some status information back in the reply to the ENABLESERVICE request.

Or, better, if possible: capture and properly hand status events that are reported between SERVICESTATUS and its reply. That might not even require changing the protocol.